scene/gui/graph_element.cpp

```diff
@@ -0,0 +1,244 @@
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "graph_element.h"
+
+#include "core/string/translation.h"
+#include "scene/gui/graph_edit.h"
+
+#ifdef TOOLS_ENABLED
+void GraphElement::_edit_set_position(const Point2 &p_position) {
+	GraphEdit *graph = Object::cast_to<GraphEdit>(get_parent());
+	if (graph) {
+		Point2 offset = (p_position + graph->get_scroll_offset()) * graph->get_zoom();
+		set_position_offset(offset);
+	}
+	set_position(p_position);
+}
+#endif
+
+void GraphElement::_close_requested() {
+	// Send focus to parent.
+	get_parent_control()->grab_focus();
+	emit_signal(SNAME("close_request"));
+}
+
+void GraphElement::_resort() {
+	Size2 size = get_size();
+
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *child = Object::cast_to<Control>(get_child(i));
+		if (!child || !child->is_visible_in_tree()) {
+			continue;
+		}
+		if (child->is_set_as_top_level()) {
+			continue;
+		}
+
+		fit_child_in_rect(child, Rect2(Point2(), size));
+	}
+}
+
+Size2 GraphElement::get_minimum_size() const {
+	Size2 minsize;
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *child = Object::cast_to<Control>(get_child(i));
+		if (!child) {
+			continue;
+		}
+		if (child->is_set_as_top_level()) {
+			continue;
+		}
+
+		Size2i size = child->get_combined_minimum_size();
+
+		minsize.x = MAX(minsize.x, size.x);
+		minsize.y += MAX(minsize.y, size.y);
+	}
+
+	return minsize;
+}
+
+void GraphElement::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_SORT_CHILDREN: {
+			_resort();
+		} break;
+
+		case NOTIFICATION_LAYOUT_DIRECTION_CHANGED:
+		case NOTIFICATION_TRANSLATION_CHANGED:
+		case NOTIFICATION_THEME_CHANGED: {
+			update_minimum_size();
+			queue_redraw();
+		} break;
+	}
+}
+
+void GraphElement::_validate_property(PropertyInfo &p_property) const {
+	Control::_validate_property(p_property);
+	GraphEdit *graph = Object::cast_to<GraphEdit>(get_parent());
+	if (graph) {
+		if (p_property.name == "position") {
+			p_property.usage |= PROPERTY_USAGE_READ_ONLY;
+		}
+	}
+}
+
+void GraphElement::set_position_offset(const Vector2 &p_offset) {
+	position_offset = p_offset;
```

Comment:
This used to have a check and return early without any assignment if the new value is the same. I think we should keep it, since the method has side-effects.